staging/src/k8s.io/client-go/rest/request.go

```diff
@@ -910,61 +913,46 @@ func (r *Request) request(ctx context.Context, fn func(*http.Request, *http.Resp
 		} else {
 			r.backoff.UpdateBackoff(r.URL(), err, resp.StatusCode)
 		}
-		if err != nil {
-			// "Connection reset by peer" or "apiserver is shutting down" are usually a transient errors.
-			// Thus in case of "GET" operations, we simply retry it.
-			// We are not automatically retrying "write" operations, as
-			// they are not idempotent.
-			if r.verb != "GET" {
-				return err
-			}
-			// For connection errors and apiserver shutdown errors retry.
-			if net.IsConnectionReset(err) || net.IsProbableEOF(err) {
-				// For the purpose of retry, we set the artificial "retry-after" response.
-				// TODO: Should we clean the original response if it exists?
-				resp = &http.Response{
-					StatusCode: http.StatusInternalServerError,
-					Header:     http.Header{"Retry-After": []string{"1"}},
-					Body:       ioutil.NopCloser(bytes.NewReader([]byte{})),
+
+		done := func() bool {
+			defer readAndCloseResponseBody(resp)
+
+			// if the the server returns an error in err, the response will be nil.
+			f := func(req *http.Request, resp *http.Response) {
+				if resp == nil {
+					return
 				}
-			} else {
-				return err
+				fn(req, resp)
 			}
-		}
 
-		done := func() bool {
-			// Ensure the response body is fully read and closed
-			// before we reconnect, so that we reuse the same TCP
-			// connection.
-			defer func() {
-				const maxBodySlurpSize = 2 << 10
-				if resp.ContentLength <= maxBodySlurpSize {
-					io.Copy(ioutil.Discard, &io.LimitedReader{R: resp.Body, N: maxBodySlurpSize})
+			var retry bool
+			retryAfter, retry = r.retry.NextRetry(req, resp, err, func(req *http.Request, err error) bool {
+				// "Connection reset by peer" or "apiserver is shutting down" are usually a transient errors.
+				// Thus in case of "GET" operations, we simply retry it.
+				// We are not automatically retrying "write" operations, as they are not idempotent.
+				if r.verb != "GET" {
```

Comment:
This change looks like it broke the previous behavior.  Prior to this PR, the following things were retried:

* "transient errors on GET"
* "explicit retriable errors from the server for any method where the body was reseekable"

After this change, we only retry

* "GET methods where a transient or explicit reply methods"

For that reason this change is incorrect and broke production behavior.  The whole change (and any dependent changes) must be reverted and we will reintroduce the new code (new factoring) once new test cases are introduced.

Reply:
I.e. the previous method:

```
		if err != nil {
			// "Connection reset by peer" or "apiserver is shutting down" are usually a transient errors.
			// Thus in case of "GET" operations, we simply retry it.
			// We are not automatically retrying "write" operations, as
			// they are not idempotent.
			if r.verb != "GET" {
				return err
			}
			// For connection errors and apiserver shutdown errors retry.
			if net.IsConnectionReset(err) || net.IsProbableEOF(err) {
				// For the purpose of retry, we set the artificial "retry-after" response.
				// TODO: Should we clean the original response if it exists?
				resp = &http.Response{
					StatusCode: http.StatusInternalServerError,
					Header:     http.Header{"Retry-After": []string{"1"}},
					Body:       ioutil.NopCloser(bytes.NewReader([]byte{})),
				}
			} else {
				return err
			}
```
^ retries transient errors

```
		}

		done := func() bool {
			// Ensure the response body is fully read and closed
			// before we reconnect, so that we reuse the same TCP
			// connection.
			defer func() {
				const maxBodySlurpSize = 2 << 10
				if resp.ContentLength <= maxBodySlurpSize {
					io.Copy(ioutil.Discard, &io.LimitedReader{R: resp.Body, N: maxBodySlurpSize})
				}
				resp.Body.Close()
			}()

			retries++
			if seconds, wait := checkWait(resp); wait && retries <= r.maxRetries {
```

^ if the error from the server indicates a retry (retry-after means the client is free to retry, on any verb), then this method will return it, which means the new logic prevents retries of POST/PUT/DELETE with bodies.

```

				retryInfo = getRetryReason(retries, seconds, resp, err)
				if seeker, ok := r.body.(io.Seeker); ok && r.body != nil {
					_, err := seeker.Seek(0, 0)
					if err != nil {
						klog.V(4).Infof("Could not retry request, can't Seek() back to beginning of body for %T", r.body)
						fn(req, resp)
						return true
					}
				}

				klog.V(4).Infof("Got a Retry-After %ds response for attempt %d to %v", seconds, retries, url)
				r.backoff.Sleep(time.Duration(seconds) * time.Second)
				return false
			}
			fn(req, resp)
			return true
		}()
		if done {
			return nil
		}
	}
```

Reply:
OK yeah - I agree with this.
Thanks for catching this Clayton

@tkashem - I agree with Clayton - at this point of the release we should revert this PR.

Reply:
> explicit retriable errors from the server for any method where the body was reseekable

are you referring to the following scenario?
- request verb is `POST`
- request body is re-seekable
- server sends `429`, `Retry-After: 1`

This should be retried - https://github.com/kubernetes/kubernetes/blob/1ea07d482a5ec15afac5ea837585174579fc0f57/staging/src/k8s.io/client-go/rest/with_retry.go#L173-L195

also, `TestCheckRetryHandles429And5xx` seems to exercise `POST` with bodies
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/rest/request_test.go#L1446-L1483

Reply:
> I agree with Clayton - at this point of the release we should revert this PR.

@wojtek-t @smarterclayton yeah, i am working on a revert PR with all dependencies.

Reply:
This is the output from `TestCheckRetryHandles429And5xx` that issues a `POST` with non-empty request body, the log shows retries successful resetting of the request body.

```
go test -race -v -count=1 k8s.io/client-go/rest -run=TestCheckRetryHandles429And5xx
=== RUN   TestCheckRetryHandles429And5xx
    request_test.go:1457: attempt 0, request verb: POST
I0330 07:42:34.722020  742601 with_retry.go:224] request body reset successful
    request_test.go:1457: attempt 1, request verb: POST
I0330 07:42:34.723030  742601 with_retry.go:224] request body reset successful
    request_test.go:1457: attempt 2, request verb: POST
I0330 07:42:34.724298  742601 with_retry.go:224] request body reset successful
    request_test.go:1457: attempt 3, request verb: POST
I0330 07:42:34.725270  742601 with_retry.go:224] request body reset successful
    request_test.go:1457: attempt 4, request verb: POST
--- PASS: TestCheckRetryHandles429And5xx (0.01s)
PASS
ok  	k8s.io/client-go/rest	0.073s
```

Reply:
two options:
- if we still want to revert the entire refactor of retry: https://github.com/kubernetes/kubernetes/pull/109149

- make `retry` thread safe in `Request` and revert the wrapping of error from a previous retry: https://github.com/kubernetes/kubernetes/pull/109148 and https://github.com/kubernetes/kubernetes/pull/109114

Reply:
@smarterclayton let me know what option you want to go with, i will close the undesired PR(s) based on your decision.

Reply:
In your test, I added some debugging to the `isErrRetryableFunc` closure in `request()` and it is not being called where I thought it was being called.

Digging deeper, it's because IsNextRetry is being called on the error from the http client, but NOT the response from the server (which is only `checkWait`).  Comparing that to 1.21 yields the same logic - we retry if `Retry-After` is present and the body is seekable, and `IsErrRetryable`.  

So you're right, this didn't change logic, and we did not regress retries when the server tells us to retry, and we have a test that proves it that is working correctly (good!).

Given that, we're back to the thread safety change being the only issue.  That is not a release blocker, but the other issue that depends on the construction of the client potentially is, which means we need to understand the scope of the change to avoid breaking thread safety (the refactor mentioned elsewhere) to figure out the risk of fixing both issues or just fixing #108906 and dealing with the thread safety afterwards.

Reply:
As a follow up for the next release, the comments on some of these methods and sections could be improved.  For instance, we may want to rename `IsErrorRetryable` to `IsHttpErrRetryable` and `checkWait` to `isRetryFromServerResponse`

Reply:
thanks for the in depth review, I closed the revert PRs, and opened a new issue with your suggestions.